lib/Sema/MiscDiagnostics.cpp

```diff
@@ -1680,34 +1656,174 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
       if (!ty->hasReferenceSemantics())
         return false;
 
-      if (auto closureExpr = dyn_cast<ClosureExpr>(inClosure)) {
-        if (auto selfDecl = closureExpr->getCapturedSelfDecl()) {
-          // If this capture is using the name `self` actually referring
-          // to some other variable (e.g. with `[self = "hello"]`)
-          // then implicit self is not allowed.
-          if (!selfDecl->isSelfParamCapture()) {
-            return true;
-          }
+      return !selfDeclAllowsImplicitSelf(DRE->getDecl(), inClosure);
+    }
+
+    static bool selfDeclAllowsImplicitSelf(const ValueDecl *selfDecl,
+                                           const AbstractClosureExpr *inClosure,
+                                           bool validateParentClosures = true) {
+      ASTContext &ctx = inClosure->getASTContext();
+
+      auto requiresSelfQualification =
+          isClosureRequiringSelfQualification(inClosure, ctx);
+
+      // If this is an explicit `weak self` capture, then implicit self is
+      // allowed once the closure's self param is unwrapped in an valid way
+      // (e.g. `guard let self` or `if let self`).
+      if (closureHasWeakSelfCapture(inClosure)) {
+        if (!implicitWeakSelfReferenceIsValid(selfDecl, inClosure)) {
+          return false;
         }
       }
 
-      if (auto var = dyn_cast<VarDecl>(DRE->getDecl())) {
-        if (!isEnclosingSelfReference(var, inClosure)) {
+      else if (auto closureExpr = dyn_cast<ClosureExpr>(inClosure)) {
+        auto capturedSelfDecl = closureExpr->getCapturedSelfDecl();
+
+        // If this closure doesn't capture self explicitly, but this closure
+        // requires self qualification, then implicit self is disallowed.
+        if (!capturedSelfDecl && requiresSelfQualification) {
+          return false;
+        }
+
+        // If the closure has an explicit capture using the name `self` that
+        // actually refers to some other variable (e.g. `[self = "hello"]`)
+        // then implicit self is not allowed.
+        if (capturedSelfDecl && !capturedSelfDecl->isSelfParamCapture()) {
```

Comment:
Additionally, it seems like this is currently rejected:

```swift
class C {
  func foo() {
    let _ = { [weak self] in
      let _ = { [self] in
        guard let self else { return }
        foo()
      }
    }
  }
}
```

Reply:
> Do we need to be checking this for the `weak` case too? Seems like we currently allow e.g:
> 
> ```swift
> class C {
>   func bar() {}
>   func foo(_ x: C) {
>     let fn = { [weak self = x] in
>       guard let self else { return }
>       bar()
>     }
>   }
> }
> ```

Ahhh thanks for noticing this! You're right that this is unexpectedly permitted in Swift 5.10. I'm surprised we missed this in the original implementation of SE-0365. 

I added a check for this, so now we reject it as expected. Since this is source breaking, it only emits a warning in Swift 5 mode, and emits an error in Swift 6 mode.

> Additionally, it seems like this is currently rejected:
> 
> ```swift
> class C {
>   func foo() {
>     let _ = { [weak self] in
>       let _ = { [self] in
>         guard let self else { return }
>         foo()
>       }
>     }
>   }
> }
> ```

This sort of case isn't explicitly covered by the text of SE-0365. All of the examples of supported patterns put the `guard let self else { return }` condition in the outer closure rather than the inner closure. 

I could see it making sense to support this, but since it isn't discussed in the proposal I think it's best to leave it as-is for now. In the current change on this branch we still reject it, like today in Swift 5.10.

Reply:
Fair enough, I'm happy to leave that as future work